ArtOfIllusion/src/artofillusion/view/ClickedPointFinder.java

```diff
@@ -36,52 +35,48 @@ public ClickedPointFinder()
     
      This works only for objects, that can produce a RenderingMesh.
    */
-   
-  public Vec3 newPoint(ViewerCanvas v, Point point)
+
+  public Vec3 newPoint(ViewerCanvas view, Point point)
   {
-    Vec3 clickedPoint = v.getCamera().convertScreenToWorld(point, v.getDistToPlane()); // defaultpoint if nothing is there
+    Vec3 clickedPoint = view.getCamera().convertScreenToWorld(point, view.getDistToPlane()); // defaultpoint if nothing is there
     Vec3 pointOnTriangle;
     boolean inSpace = true;
     bary = new double[3];
 
-    cameraOrigin = v.getCamera().getCameraCoordinates().getOrigin();
-    cameraZ = v.getCamera().getCameraCoordinates().getZDirection();
-    perspective = v.isPerspective();
-    w = v.getBounds().width;
-    h = v.getBounds().height;
-  
+    cameraOrigin = view.getCamera().getCameraCoordinates().getOrigin();
+    cameraZ = view.getCamera().getCameraCoordinates().getZDirection();
+    perspective = view.isPerspective();
+    w = view.getBounds().width;
+    h = view.getBounds().height;
+
     // ToScreen matrices produce the actual pixel coordinates on the ViewerCanvas.
-    
-    if ((v instanceof ObjectViewer) && !(((ObjectViewer)v).getUseWorldCoords()))
-      modelToScreen = v.getCamera().getObjectToScreen();
+
+    if ((view instanceof ObjectViewer) && !(((ObjectViewer)view).getUseWorldCoords()))
+      modelToScreen = view.getCamera().getObjectToScreen();
     else
-      modelToScreen = v.getCamera().getWorldToScreen();
+      modelToScreen = view.getCamera().getWorldToScreen();
     
-    Mat4 toScene, toThisObject, fromExtToLocal;
-    ObjectInfo[] objList = renderableObjects(v);
-    RenderingMesh rMesh;
+    RenderingMesh surface;
+    boolean hideFace[];
     Vec3[] corner3D = new Vec3[3];
     Vec2[] corner2D = new Vec2[3],  corner2DS = new Vec2[3];
-    CoordinateSystem localCoords;
+    Mat4 toContext;
+    ArrayList<ObjectInfo> shownObjects = renderableObjects(view);
 
-    localCoords = getLocalCoords(v);
-    toThisObject = localCoords.toLocal();
-
-    for (int i = 0; i < objList.length ; i++)
+    for (ObjectInfo info : shownObjects)
     {
-      rMesh = objList[i].getPreviewMesh();
-      toScene = objList[i].getCoords().fromLocal();
-      fromExtToLocal = toThisObject.times(toScene);
-      hideFace = v.getHiddenFaces();
+      surface = info.getPreviewMesh();
+      hideFace = view.getHiddenFaces();
+      toContext = contextTransform(view, info);
 
-      for (int j = 0; j < rMesh.triangle.length; j++)
+      for (int f = 0; f < surface.triangle.length; f++)
       {
-        if (hideFace != null && hideFace[j])
+        if (hideFace != null && hideFace[f])
```

Comment:
I figured out where my exceptions were coming from.

This code assumes that `hideFace` is the same length as `surface`, if it exists at all. However, `surface` is a `RenderingMesh`, and `hideFace` is keyed on the control mesh. They only match if the mesh is in 'Smoothing:none' or 'Smoothing:shaded' mode. (Or if _all_ of the edges and vertices are 'Smoothness:0.0')

When they _don't_ match, `surface` will definitely be longer, and this line will throw `ArrayIndexOutOfBoundsException` and the click-to-center will do nothing until you change mesh mode, or show the entire mesh.

Probably want to use the control mesh instead.

Reply:
Hmm... The canvas dawers get a rendering mesh and the hidden faces list. I'll check again how they are used there.

Reply:
So, `hideFace[]` in a CanvasDrawer is not the same `hideFace[]` that appears in `TriMeshEditorWindow`

The correct hide list was produced by `drawSurface()`in `TriMeshViewer`. All I had to do was to lift it from the method to class level and use that one. Renamed the relevant parts.